### PR TITLE
Adjoint dft

### PR DESCRIPTION
**Pull request by landmanbester**
This was driving me crazy. @sjperkins can you please have a look at the code in the dft sub-directory? For some reason the tests fail (by hundreds of orders of magnitude) when I initialise the arrays in kernels with np.empty. Not sure if this is necessary for the dask version...

### CODE REVIEW

**Comment by sjperkins** [comment]
@landmanbester  Your kernels have these expressions in them

```python
vis_of_im[row, chan] += np.exp(p)*image[source, chan]
```

Since `vis_of_im[row,chan]` isn't initialised when using `np.empty` I don't think you get the expected values. Maybe `np.zeros` is the way to go, unless you can think of a way to do the initialisation in the numba code.

Also be sure to add

```python
# -*- coding: utf-8 -*-

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```

at the top of your files as this will fix most of the python 3 failures.

**Comment by landmanbester** [comment]
@sjperkins think we can merge this in? I am pretty happy with it for now

File: africanus/dft/dask.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        def _wrapper(image, uvw, lm, frequency, dtype_):
+            return np_im_to_vis(image[0], uvw[0], lm[0][0],
+                                frequency, dtype=dtype_)
+        assert lm.chunks[0][0] == lm.shape[0], \
+            "lm chunks must match lm shape on first axis"
+        assert image.chunks[0][0] == image.shape[0], \
+            "Image chunks must match image shape on first axis"
+        assert image.chunks[0][0] == lm.chunks[0][0], \
+            "Image chunks and lm chunks must match on first axis"
+        return da.core.atop(_wrapper, ("row", "chan"),
```
Comment (sjperkins) [inline africanus/dft/dask.py]: I prefer:

```python
if cond:
  raise ValueError("...")
```

File: tests/test_dft.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    lm_dask = da.from_array(lm, chunks=(5, 2))
+    frequency_dask = da.from_array(frequency, chunks=4)
+    vis_dask = da.from_array(vis, chunks=(25, 4))
+
+    image_dask = dask_vis_to_im(vis_dask, uvw_dask, lm_dask, frequency_dask)
+
+    assert np.allclose(image, image_dask)
+
+# if __name__=="__main__":
+#     test_im_to_vis()
```
Comment (sjperkins) [inline tests/test_dft.py]: Remove comments

File: tests/test_dft.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+import numpy as np
+
+import pytest
+
+def test_im_to_vis():
+    """
+    The simplest test here is to see if a single source at the phase centre
+    returns simply the flux everywhere with zero imaginary part
+    :return:
```
Comment (sjperkins) [inline tests/test_dft.py]: Remove `:return:`

File: tests/test_dft.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    for i in range(nchan):
+        assert np.all(image[:, i] == wsum)
+
+
+def test_adjointness():
+    """
+    She is the mother of all tests. The DFT should be perfectly self adjoint up to 
+    machine precision. 
+    :return:
```
Comment (sjperkins) [inline tests/test_dft.py]: Remove `:return:`

File: tests/test_dft.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        assert np.all(tmp.real < 1e-13)
+        assert np.all(tmp.imag < 1e-13)
+
+
+def test_vis_to_im():
+    """
+    Still thinking of a better test here but we can do here but the simplest test 
+    does exactly the same as the above. If we have an auto-correlation we expect 
+    to measure a flat image with value wsum
+    :return:
```
Comment (sjperkins) [inline tests/test_dft.py]: Remove `:return:`

File: africanus/dft/kernels.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    return _vis_to_im_impl(vis, uvw, lm, frequency, im_of_vis)
+
+
+_DFT_DOCSTRING = namedtuple(
+    "_DFTDOCSTRING", ["preamble", "parameters", "returns"])
+
+im_to_vis_docs = _DFT_DOCSTRING(
+    preamble="""
+    Computes the discrete image to visibility mapping of an ideal 
+    unpolarised iterferometer :
```
Comment (sjperkins) [inline africanus/dft/kernels.py]: interferometer

**Comment by sjperkins** [comment]
thanks for the PR @landmanbester

**Pull request merged**